File: Same Tree.py
# Link: https://leetcode.com/problems/same-tree/submissions/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    
    def bfs(self, p: TreeNode, q: TreeNode):
        
        # If either is a NoneType, return false
        if (p != None and q == None) or (p == None and q != None):
            return False
        
        # If both are none, return true
        if p == None and q == None:
            return True
        
        # Return false, if both nodes don't share the same value
        if p.val != q.val:
            return False
        
        # Check all children
        leftChildCheck = self.bfs(p.left, q.left)
        rightChildCheck = self.bfs(p.right, q.right)
        
        # If either variable is false, return false
        if leftChildCheck == False or rightChildCheck == False:
            return False
        
        return True
    
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        
        return self.bfs(p, q)

File: Same Tree.java
// https://leetcode.com/problems/same-tree/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {

        /*
         * Time Complexity: O(n) where n = the number of nodes in the largest tree. During the recursive call,
         *  all nodes will be called upon and inspected. Even if one tree has less nodes than the other, the 
         *  algorithm will still check the tree with the most nodes.
         * 
         * Space Complexity: O(h) where h = the tree with the tallest height. This is a recursive function, so
         *  for every new method call, a stack frame will be added. The depth of the tree will influence how many
         *  stack frames will be created.
         */
        
        // If both are null, then trees are identical
        if (p == null && q == null) {
            return true;
        }

        // If either root is null, then trees aren't identical
        if ((p == null && q != null) || (p != null && q == null)) {
            return false;
        }

        // Check if roots share the same value
        if (p.val == q.val) {

            // Check if both subtrees are identical
            boolean leftSubtree = isSameTree(p.left, q.left);
            boolean rightSubtree = isSameTree(p.right, q.right);

            return leftSubtree && rightSubtree;
        }

        return false;
    }
}